README.txt
pts-quote: display a random quote on the console
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pts-quote (PotterSoftware Quote Displayer) is a collection of command-line
tools for choosing and displaying a text quote (tagline, quote-of-the-day,
QotD) randomly from a text file and displaying it on the console. Currently
it is implemented in assembly language for DOS on the Intel 8086 and 286
architectures (and it can be run in DOSBox and QEMU on modern systems).

The inital assembly language implementation for DOS on the Intel 286
architecture was written in 1996 for Turbo Pascal 7.0 inline assembly and
the A86 3.22 assembler, with the last 1996 version being quote6.com.orig in
this repository. Maintenance work (mostly software archeology, porting to
the NASM 0.98.39 and Yasm 1.2.0 assemblers and bugfixes) resumed on
2022-03-18. See the file changelog.txt for more details.

How to try and run the latest version:

* Download the latest snapshot of the pts-quote repository.
* Copy quote63.com.orig to quote.com .
  To try the latest old version from 1996 instead, copy quote5.com.orig
  instead to quote.com .
* Copy quote_demo_dos.txt to quote.txt .
* Install DOSBox.
* Run `dosbox .' (without the quotes) in the repository directory.
* Within the DOSBox window, run: del quote.idx
* Within the DOSBox window, run: quote
  This should display a random quote with a colorful banner and a frame.
  Run it again to get a different random quote.

How to compile and build:

* Most users don't need to compile from source: there are many precompiled
  .com.orig and .exe.orig files in the pts-quote repository. Just copy one
  of the .com.orig files to quote.com or one of the .exe.orig files to
  quotex.exe, and run quote.com or quotex.exe on DOS or in a DOS emulator.
* Each .pas file is a Turbo Pascal 7.0 source file, and contains compile
  instructions near the beginning (to create a DOS .exe program). To
  compile, you need the Turbo Pascal 7.0 compiler (tpc.exe and its data
  files), and you have to run it on DOS or in a DOS emulator (e.g. DOSBox).
* Each .8 file is an assembly source file in the A86 dialect (works with A86
  versions 3.22 .. 4.05), and contains compile instructions near the
  beginning (to create a DOS .com program). To compile, you need the A86
  assembler (latest version 4.05 available on http://eji.com/a86.zip ), and
  you have to run it on DOS or in a DOS emulator (e.g. DOSBox).
* Each .nasm file is an assembly source file in the NASM dialect (works with
  NASM versions 0.98.39 .. 2.13.02 ... and Yasm versions 1.2.0 and 1.2.0,
  and possibly other versions of NASM and Yasm), and contains compile
  instructions near the beginning (to create a DOS .com program). To compile,
  you need the NASM or Yasm assembler (both available for free as packages
  for Linux and macOS, and executable programs available for Windows). You
  can run these assemblers directly on modern (2022) hardware, without the
  need for emulation. However, to run the compiled DOS .com program, you
  need a DOS emulator (such as DOSBox, see instructions above) or a very
  old system running DOS.

File format of the text file quote.txt as of version 2.63:

* Quotes are separated by one or more empty lines.
* Lines are separated by LF or CRLF (can be mixed). At the end of the
  file there may be 0 or more LF or CRLF.
* Each quote must fit to 4095 bytes (including the trailing empty line).
* If a line starts with --, it will be right-aligned without the -- ,
  and it will be highlighted.
* If a line starts with -&, it will be center-aligned without the -& ,
  and it will be highlighted.
* Otherwise lines must not start with - .
* The file quote.txt must be at most 61.25 MiB in size.
* The file quote.txt must contain at least 1 quote.
* Each 1024-byte block of quote.txt must contain at most 255 quotes. That
  can be ensured by making each quote at least 5 bytes long (including the
  trailing empty line).
* The maximum number of quotes in quote.txt is 15 993 600. This is because
  quote.idx can describe up to 62720 1024-byte blocks, with up to 255 quotes
  each.
* Each quote line must be at most 78 bytes (characters) long, excluding the
  LF or CRLF terminator. This is only verified when printing the individual
  quote.
* To check whether quote.txt contains any errors (except for line-too-long
  and quote-too-long), i.e. to validate it, run quote63n.com with parameter
  "i" (without the quotes). This will also regenerate the index file
  quote.idx as a side effect. There are errors iff a line starting with E is
  printed.

Further restrictions on quote.txt imposed by some earlier versions:

* Each line including the last one must be terminated with a CRLF (\r\n,
  ASCII 10 + 13).
* Quotes are separated by an empty line (i.e. CRLF + CRLF), and the last
  line must also be empty.
* If the last empty line (i.e. CRLF + CRLF) is missing, then then versions
  before 2.60 will ignore such a last quote, versions 2.60 .. 2.62 will exit
  early without finishing.
* Quotes must not be empty (i.e. CRLF + CRLF must not be followed by CRLF).
* Each quote must fit to 4095 bytes (including the trailing CRLF + CRLF).
* The file quote.txt must be at most 55 MiB in size in versions <=2.62.
* The file quote.txt must contain at most 24160 quotes in versions <2.60.
* The file quote.txt must contain at most 65535 quotes in version 2.62.

About the index file quote.idx:

* The index file is autogenerated from quote.txt if it doesn't exist.
* If you modify quote.txt, delete quote.idx, there is no autodetection
  for regeneration.
* The index file contains file offset information to make it faster to find
  the random quote in quote.txt. The index file is binary.
* The file format of the index file depends on the version of the program.
  Versions 2.30 .. 2.5? use the verbose format 5, versions since 2.60 use
  the compact format 6. Delete quote.idx between using different versions.
* Format 5 (2..30 .. 2.5?) of quote.idx index file:
  * For each quote, encode its byte size (including the trailing CRLF +
    CRLF) as 1 or 2 8-bit unsigned integers. If the size is at most 239,
    emit it as 1 8-bit unsigned integer, otherwise emit 240 + the high 4
    bits as the first 8-bit unsigned integer, and then emit the low 8 bits
    as the second 8-bit unsigned integer. Equivalently, for decoding: read
    an 8-bit unsigned integer to hi. If it's at most 239, use it as the byte
    size of the next quote, otherwise read another 8-bit unsigned integer to
    hi, and use ((hi - 240) * 256 + lo) as the byte size of the next quote.
* Format 6 (2.6? ..) of quote.idx index file:
  * Emit the total number of quotes, as 2 bytes containing an 16-bit
    unsigned little-endian integer.
  * For each block of 1024 bytes (the last block may be shorter) in
    quote.txt, emit the number of quotes whose first byte this block
    contains, as a byte containing an 8-bit unsigned integer.
  * An invariant: the first 2 bytes (as a 16-bit unsigned little-endian
    integer) equals to the sum of the subsequent bytes (as 8-bit unsigned
    integers).

pts-quote is free software released under the GNU GPL v2 license. There is
NO WARRANTY. Use at your own risk.

About the choice of programming language:

* The goal was to create working, convenient and fast program ready for
  everyday use, and also to impress the user with the tiny size of the
  executable program.
* The source code in 1996 was written in Turbo Pascal 7.0 (.pas files) and
  A86 3.22 (.8 files), because the author was familiar with those at the
  time. More specifically, the first version was written in Turbo Pascal 7.0
  (source code lost by now), then it was ported to Turbo Pascal 7.0 inline
  assembly (see quote3.pas in the pts-quote repository), then it was ported
  to A86 assembly (see quote6.8) to make the executable program size smaller
  by removing the .exe header and the Turbo Pascal 7.0 runtime library.
* For the 2022 updates, some the A86 source code (.8 files) was ported to
  NASM (.nasm files), because of the simplicity and versatilitiy of the NASM
  assembly languange and the good native availability of free assemblers
  (NASM and Yasm) on modern systems. The source code works with old and new
  versions of NASM: the earliest supported one is NASM 0.98.39 (released on
  2005-01-15).
* NASM output (with `nasm -O0 -f bin') is compact and deterministic, and
  with NASM it's possible (and easy most of the time) to generate
  byte-by-byte identical executable program output, matching the output of
  A86 and earlier NASM versions. This is great for reproducible builds.
* Most other assemblers (such as MASM, TASM, LZASM, Wasm, JWasm, GNU AS)
  except for FASM can't produce a DOS .com executable program directly (but
  they produce an object file, from which the linker produces an
  executable). This would introduce complexity and inconvenience to the
  build and development process.

__END__


changelog.txt
-*- coding: utf-8 -*-

History of the quote database (QUOTE.TXT):

* 2.41 QUOTE.TXT expanded. It's now 1400K long. (Way below the upper limit.)
* 2.43 Added quotes from Taylor.

History of old releases (now lost):

* 0.0  J.C.R.'s VRQ program (I decided to write a same but better program. )
* 0.1  MOTTO.*
* 1.0  Indexer routine
* 1.1  Plus Random Displayer (displays characters)
* 1.2  Line Displayer
* 1.25 Is able to align lines
* 1.3  Plus Header & Footer lines
* 1.4  Support of ANSI.SYS (if detected), bright quotes
* 2.00 Command line handling & documentation & RH
* 2.10 Optimized & reduced code (the core routines are rewritten in assembly)
* 2.20 Compressed Indextable.
* 2.30 16K preread cache for fast indexing; QUOTE.TXT can be up to 24000 lines.
* 2.31 Annoying bugs fixed.
* 2.32 Code has been optimized. (4500 bytes)

History of code written in 1996 and updated in 2022:

* 2.33 Rest of code rewritten in Turbo Pascal 7.0 internal assembly.
       Random number generation is very slow. May contain a bug.
       quote3.pas is Turbo Pascal 7.0 source code with just inline assembly.
       quote3.exe.orig is 2848 bytes.
* 2.34 From 2.33 with random number generator replaced with a faster one.
       Released on 2022-03-27.
       quote34.pas is Turbo Pascal 7.0 source code with just inline assembly.
       quote34.exe.orig is ~2832 bytes.
* 2.40 Turbo Pascal 7.0 inline assembly ported to A86 3.22 assembly.
       Source code and .exe program binary have been lost.
* 2.41 Turbo Pascal 7.0 inline assembly ported to NASM 0.98.39 assembly
       on 2022-03-27. Random number generation is very slow. Contains a
       bugfix related to ANSI.SYS detection.
       quote4.nasm is NASM 0.98.39 source code.
       quote4.com.orig are 1251 bytes, first built on 2022-03-27.
* 2.42 From 2.41 random number generator replaced with a faster one.
       Released on 2022-03-27.
       quote42.nasm is NASM 0.98.39 source code.
       quote42n.com.orig is ~1235 bytes, first built on 2022-03-27.
* 2.50 ANSI.SYS is no longer needed for color display.
       Released in PotterSoftware Entry Pack. Random number generation is
       very slow.
       quote5.8 is A86 3.22 source code reverse engineered from quote5.com.orig.
       quote5.com.orig is 1192 bytes.
* 2.51 .com program is even smaller.
       Source code and .com program binary of 984 bytes have been lost.
* 2.52 From 2.50 random number generator replaced with a faster one.
       Released on 2022-03-27.
       quote52.8 is A86 3.22 source code, most of it reverse engineered.
       quote52.com.orig is ~1192 bytes, first built on 2022-03-27.
* 2.60 QUOTE.IDX index file format redesigned, file is much smaller:
       just 8% of before, just 0.976% of QUOTE.TXT.
       Source code and .com program binary have been lost.
       In the header message, it is branded ``PotterSoftware Fortune Teller'',
       rather than ``PotterSoftware Quote Displayer''.
* 2.61 Fully working, smaller.
       Source code and .com program binary of 784 bytes have been lost.
* 2.62 Even smaller. Parameter- and full error handling included. Bugs fixed.
       Random number generator is fast.
       Released on 1996-09-22.
       quote6.8 is A86 3.22 .. 4.05 source code.
       quote6.nasm is NASM 0.98.39 source code.
       quote6.com.orig is 794 bytes. It was distributed as quote.com in 1996.
* 2.63 From 2.62 some bugfixes and compatibility improvements with previous
       versions. It can print the last quote even if file not terminated by
       CRLF+CRLF. It can use LF and/or CRLF as line terminator. It ignores
       any number of empty lines between quotes. It reports error codes.
       Work started on 2022-03-30. Also works on 8086 (not only 80286).
       quote63.nasm is NASM 0.98.39 source code.
       quote63.com.orig is ~1050 bytes.
* 2.64 quotex.nasm is under development since 2022-03-18, it's not released.
       TODO: Linux i386 version with ANSI codes for linux and xterm terminals
       TODO: more strict with leading -- and -&
       TODO: autodetect if the index file needs to be regenerated
       TODO: better handle quotes longer than 4096 bytes
       TODO: add quote.idx compatibility with version 2.64: keep idxc small
             so that it's at most 65535 quotes in 55 << 10 blocks
* 2.70 A DOS device driver (loadable from CONFIG.SYS) variant of 2.62,
       written on 1997-03-22. It contains a text font, and sets up a splash
       screen for booting.
       Not released to the general public, backed up.
       Source file is fquote.8, binary is fquote.bin.

__END__


quote34.pas
{$G+} {$F-} {$A-} {$D-} {$E-} {$I-} {$L-} {$S-} {$X-} { -*- coding: UTF-8 -*- }
{
quote34.pas: PotterSoftware Quote Displayer V2.34 (Turbo Pascal 7.0 inline assembly source code)
(C) 1996--2022-03-27 by EplPáj of PotterSoftware, Hungary

Compile it with Turbo Pascal 7.0 on DOS, generate quote34.exe: tpc quote34.pas

The QUOTE.IDX index file format is identical in version 2.30 .. 2.5? and
different from version 2.60.

It uses ANSI.SYS for color output, and it detects the lack of ANSI.SYS (such as
in DOSBox), and then it prints colorless output.

Command-line argument (first byte on the command-line):

* ' '-> Display a quote using index table (default).
* 'A'-> Display a quote using linear search.
* 'B'-> Create index table & then display a quote using it.
* 'C'-> Create index table.
}
program PotterSoftware_Quote_Displayer;

const
  full=16384;
  ttt: array[0..29] of char= #27'[44;30m$'#27'[0m'#27'[K$'#27'[30;1m$'#27'[0m$';
  txtfn: array[0..12] of char= 'QUOTE.TXT'#0#0#0#0;
  idxfn: array[0..12] of char= 'QUOTE.IDX'#0#0#0#0;
  headermsg: string[35]= 'PotterSoftware Quote Displayer 2.34';
  footermsg: string[44]= 'Greetings to RP,TT,FZ/S,Blala,OGY,FC,VR,JCR.';

var
  buf: array[0..full+4-1] of char;
  s: string absolute buf;
  idx: array[0..24160] of word;  { 1 less is enough, first element unused. TODO: Get rid of idx[0]. }
  qqq: record
    a,b,w: word;
    l, max, oldl: longint;
    xch: char;
    before: word;
    ansich: char;
    han: word;
  end absolute headermsg; { Must be long enough (23 bytes) }

function GetNext: char; assembler;
  asm
	cmp qqq.b, full+4                 { if qqq.b=full+4 then begin }
	jne @88

	mov si, offset buf + full  { move(buf[full], buf[0], 4); }
	mov di, offset buf
	movsw
	movsw

	{ blockread(f, buf[4], full, qqq.w); }
	mov ah, 3Fh
	mov bx, qqq.han
	mov cx, full
	mov dx, offset buf+4
	int 21h
	jnc @87
	mov ax, 4CF1h  { Abort on read error. }
	int 21h
@87:    mov qqq.b, 4                      { endif }

@88:    mov bx, qqq.b			{ GetNext:=Buf[qqq.B];}
	mov al, byte ptr buf[bx]
	inc qqq.b				{ inc(qqq.b); }
	add word ptr qqq.l,1		{ inc(qqq.l); }
	adc word ptr qqq.l[2], 0
end;

procedure Header(const s: OpenString); assembler;
  asm
        cmp byte ptr ttt, '*'  { Check ansi driver }
        je @71
        mov dx, offset ttt[25] { Reset every attributes first }
        mov ah, 9
        int 21h
        mov dx, offset ttt[0]
        mov ah, 9
	int 21h
@71:    mov al, 0b2h  { '▓' }
	mov cx, 3
@72:    int 29h; int 29h; int 29h; int 29h; int 29h
        dec al
        loop @72
	push ds
        lds si, s
        lodsb
	mov ah, 0
        mov dx, ax
	shr ax, 1
        mov cx, 25
        sub cx, ax
        mov bx, cx
        mov al, ' '
	jcxz @74
@75:    int 29h
	loop @75
@74:    mov cx, [ds:si-1]
	mov ch, 0
@76:    lodsb
	int 29h
        loop @76
        pop ds
        mov cx, 50
	sub cx, bx
        sub cx, dx
        mov al, ' '
	jcxz @78
@77:    int 29h
	loop @77
@78:    mov al, 0b0h  { '░' }
	mov cx, 3
@73:    int 29h; int 29h; int 29h; int 29h; int 29h
	inc al
	loop @73
	cmp byte ptr ttt, '*'
	je @79
	mov dx, offset ttt[9]
	mov ah, 9
	int 21h
@79:
end;

procedure PrintLine(w: word); assembler;
  asm
	mov al, byte ptr w
        int 29h
	mov cx, 78
	mov al, 0c4h  { '─' }
@70:    int 29h
	loop @70
	mov al, byte ptr w[1]
	int 29h
end;

label llc,lld,lle,llf,lls,after_random,fatal_error;

begin { Főprogram }
  asm { This statement is going to be long. Very long. }
	mov al, 13				{ Writeln }
	int 29h
	mov al, 10
	int 29h

	{ Detect ANSI.SYS. }
	xor ax, ax { ansi:=memw[0:$29*4+2]>memw[0:$20*4+2]; }
	mov es, ax
	mov bx, [es:$29+$29+$29+$29+2]
	cmp bx, [es:$20+$20+$20+$20+2]
	jae @95
	mov byte ptr ttt, '*' { This means there's no ANSI.SYS }
@95:

	push ds
	push offset headermsg
	call Header  { Print header. }

	{ Get parameter from beginning of command-line arguments. }
	mov ax, PrefixSeg  { xch:=char(mem[PrefixSeg:$81]); }
	mov es, ax
	mov al, [es:81h]
	cmp al, ' '
	jne @96  { if xch=' ' then xch:=char(mem[PrefixSeg:$82]); }
	mov al, [es:82h]  { First byte of command-line arguments. }
@96:    cmp byte ptr [es:80h], 0 { if mem[PrefixSeg:$80]=0 then xch:=' '; }
	jne @94
	mov al, ' '
@94:    and al, 255-32
	mov qqq.xch, al
	mov ax, ds
	{ es will remain this way for the rest of the run.
	  It affects es:, stosb, stosw, scasb, scasw, movsb, movsw, cmpsb, cmpsw. }
	mov es, ax

	{ qqq.w:=XReset(IdxFn); }
	mov ax, 3D00h { Open for Read Only, C-Mode }
	mov dx, offset idxfn
	int 21h
	mov qqq.han, ax
	sbb ax, ax  { AX:=0 if ok; AX:=$FFFF on error. }
	mov word ptr idx, 0			{ idx[0]:=0 }
	cmp ax, 0                           { if (IOResult<>0) or (xch<>#0) then }
	jne @90
	cmp qqq.xch, 0
	jne @90
	jmp lls
@90:    { XReset(TXTFN); }
	mov ax, 3D00h { Open for Read Only, C-Mode }
	mov dx, offset txtfn
	int 21h
	jc fatal_error
	mov qqq.han, ax

	{  qqq.max:=filesize(f); }
	mov ax, 4202h
	mov bx, qqq.han
	xor cx, cx
        xor dx, dx
        int 21h
        mov word ptr qqq.max, ax
        mov word ptr qqq.max[2], dx
        mov ax, 4200h
	mov bx, qqq.han
	xor cx, cx
	xor dx, dx
	int 21h

	mov qqq.b, full+4               	{b:=full+4 }
	xor ax, ax
	mov word ptr qqq.l, ax          	{l:=0 }
	mov word ptr qqq.l[2], ax
	mov word ptr qqq.oldl, ax       	{oldl:=0 }
	mov word ptr qqq.oldl[2], ax
	mov qqq.a, 1                    	{ a:=1 }
	mov [offset buf+full], ax  { Make sure we don't detect CRLF+CRLF at the beginning. }

					{repeat }
@81:    call GetNext
	mov si, offset buf-4
	add si, qqq.b
	cmp word ptr [ds:si], $0A0D
	jne @82
	cmp word ptr [ds:si+2], $0A0D
	jne @82                            {  if buf[b-4]=newline then begin }

	mov ax, word ptr qqq.l                {    idx[a]:=l-oldl }
	mov dx, word ptr qqq.l[2]
	sub ax, word ptr qqq.oldl
	sbb dx, word ptr qqq.oldl[2]
	mov di, offset idx
	add di, qqq.a
	add di, qqq.a
	mov [ds:di], ax
	mov [ds:di+2], dx
	inc qqq.a                             {    inc(a) }
	mov ax, word ptr qqq.l                {    oldl:=l }
	mov word ptr qqq.oldl, ax
	mov ax, word ptr qqq.l[2]
	mov word ptr qqq.oldl[2], ax          {  end; }

@82:    mov ax, word ptr qqq.max              {until l=max }
	mov dx, word ptr qqq.max[2]
	cmp word ptr qqq.l, ax
	jne @81
	cmp word ptr qqq.l[2], dx
	jne @81

	{ Now qqq.a == (number of quotes) + 1. }
	{ close(f); }
	mov ah, 3Eh
	mov bx, qqq.han
	int 21h

	cmp qqq.xch, 'A'  { Don't write the index file on parameter 'A'. }
	je llc
	mov cx, qqq.a
	dec cx
	mov si, offset idx+2
	mov di, offset buf
	xor dx, dx
@83:    lodsw
	stosb
	cmp ax, 240
	jb @84
	dec di
	rol ax, 8
	or ax, 240  { 11110000b }
	stosw
@84:    loop @83
	sub di, offset buf  { DI := number of compressed bytes to write. }

        { XRewrite(IDXFN); }
        mov ah, 3Ch { Create file }
        mov cx, 0
        mov dx, offset idxfn
	int 21h
	jnc @91
fatal_error:
        mov ax, 4CF0h
	int 21h { Fatal error }
@91:    { blockwrite(f, buf, compressed_size); }
	mov ah, 40h
	mov bx, qqq.han
	mov cx, di
	mov dx, offset buf
	int 21h
	{ close(f); }
	mov ah, 3Eh
	mov bx, qqq.han
	int 21h
	{ goto c; }
	jmp llc
				{ end else begin }
lls:    { XReset(IDXFN); }
	mov ax, 3D00h { Open for Read Only, C-Mode }
	mov dx, offset idxfn
	int 21h
	jc fatal_error
	mov qqq.han, ax
	{ blockread(f, buf, full+4, reg_ax); }
	mov ah, 3Fh
	mov bx, qqq.han
	mov cx, full+4
        mov dx, offset buf
        int 21h
	mov cx, ax  { Save number of (compressed) bytes read to cx, for below. }
	{ close(f); }
	mov ah, 3Eh
	mov bx, qqq.han
	int 21h

	mov dx, 1
	mov si, offset buf
	mov di, offset idx+2
@85:    lodsb
	mov ah, 0
	stosw
	cmp al, 240
	jb @86
	dec di
	dec di
	and al, 15
	mov ah, al
	lodsb
	stosw
	dec cx
@86:    inc dx
	loop @85
	mov qqq.a, dx  { qqq.a := (number of quotes) + 1. }
llc:    cmp qqq.xch, 'C'
	je llf

	{ XReset(TXTFN); }
	mov ax, 3D00h { Open for Read Only, C-Mode }
	mov dx, offset txtfn
	int 21h
	jc fatal_error
	mov qqq.han, ax
	{ Now qqq.a-1 is the number of quotes in txtfn, provided that txtfn ends with CRLF + CRLF. }
	xor ax, ax
	mov word ptr qqq.l, ax
	mov word ptr qqq.l+2, ax
	mov si, qqq.a
	test si, si
	jz after_random  { If there are 0 quotes, print from the beginning of txtfn. }
	dec si  { SI := Number of quotes. }

	{ DX:=random(SI);  Then 0 <= SI < DX.
	  This code may ruin AX, BX, CX, SI, DI, BP and flags.

	  This code generates a 32-bit random number n in a register pair, then computes
	  DX := (n * SI) >> 32 as the random value.
	}
	push bp
	{mov ah, 0}  { Not needed, AX=0 above. }
	int 01ah  { Get time-random seed in CX:DX. }
	xor bp, bp
	mov ax, cs
	add ax, dx  { Modify seed. }
	mov bx, ax
	mov dx, 08405h
	mul dx
	shl bx, 3
	add ch, cl
	add dx, bx
	add dx, cx
	shl cx, 2
	add dx, cx
	add dh, bl
	shl cx, 5
	add ax, 1  { Modifies CF (inc ax doesn't).. }
	adc dx, bp
	mov bx, dx
	mul si
	mov ax, bx
	mov bx, dx
	mul si
	add ax, bx
	adc dx, bp  { DX:=random(SI) }
	pop bp
	jz after_random  { If the chosen random number is 0, print from the beginning of txtfn. }

	{ Calculate the start offset of quote qqq.w into qqq.l:
	  qqq.l := idx[qqq.w] + idx[qqq.w-1] + ... + idx[qqq.1]. }
        mov si, offset idx
	add si, dx
	add si, dx
	std
@97:    lodsw
	add word ptr qqq.l, ax
	adc word ptr qqq.l[2], 0
	cmp si, offset idx
	jne @97
	cld

after_random:
	push $BFDA
	call PrintLine

lld:    { seek(f, qqq.l); }
        mov ax, 4200h
	mov bx, qqq.han
        mov dx, word ptr qqq.l
        mov cx, word ptr qqq.l[2]
        int 21h
        { blockread(f, s[1], 255, qqq.w); }
        mov ah, 3Fh
        mov bx, qqq.han
        mov cx, 255
        mov dx, offset s+1
        int 21h
        mov qqq.w, ax

        cmp qqq.w, 0			{ Stop at EOF }
        je lle
        mov bx, 0			{ Look for #13 to determine length(s) }
        mov si, offset s+1
@12:    or bh, bh
        jnz @13
        cmp byte ptr [si+bx], 13
        je @11
	inc bx
        jmp @12
@13:    jmp fatal_error { Error: Line longer than 255 bytes. }
@11:    mov byte ptr s, bl { Beállítjuk a string hosszát }
        inc bx
        inc bx
        add word ptr qqq.l, bx { inc(l,length(s)+2); }
        adc word ptr qqq.l[2], 0

	{If S='' align returns TRUE else it returns FALSE. Align prints S with the
	correct color & alignment according to the control codes found in S[1,2]}

{ START OF ALIGN }

	{ Calculate the value of BEFORE first using up AnsiCh: #0=Left '-'=Right
	'&'=Center alignment }
	mov cx, ds
        mov si, offset s
        mov qqq.w, si
        lodsb
	cmp al, 0
        jne @d
        mov al, 1
        mov ds, cx
	jmp @9  { Empty string: do nothing but restore original CS. }
@d:     mov qqq.ansich, 0 { AnsiCh is 0 by default }
        cmp byte ptr [ds:si], '-'
        jne @c
        mov al, [ds:si+1]
        mov qqq.ansich, al
        add qqq.w, 2 { If not AnsiCh<>0 the 1st 2 char won't be in the str }
        mov al, [ds:si-1]
        dec ax
        dec ax
        mov [ds:si+1],al
@c:     mov ah, 0
        mov ds, cx
        mov bx, 78
        cmp qqq.ansich, 0
        jne @a
	mov al, 0
        mov bx, 0
@a:     cmp qqq.ansich, '&'
        jne @b
        mov bx, 39
        shr ax, 1
@b:     sub bx, ax
        mov qqq.before, bx
        mov al, 0b3h  { '│' } { The line starts by this }
        int 29h
        cmp byte ptr ttt, '*' { Put out an ANSI EscSeq to set color if needed }
        je @6
        mov ah, 9
	mov al, qqq.ansich
	cmp al, 0
        je @6
        add al, 10
        mov byte ptr ttt[17+3], al
        mov dx, offset ttt[17]
        int 21h
	mov qqq.ansich, 0
@6:     push ds { Display the string "s" with "before" spaces in front of it }
        mov si, qqq.w
        lodsb
        mov cl, al
        mov ch, 0
        mov dx, cx
        jcxz @1
        mov cx, qqq.before
        jcxz @5
        mov al, ' '
@2:     int 29h
        loop @2
@5:     mov cx, dx
        jcxz @8
@3:     lodsb
        int 29h
        loop @3
@8:     mov cx, 78
        sub cx, qqq.before
	sub cx, dx
	jcxz @1
	mov al, ' '
@4:     int 29h
	loop @4
@1:     pop ds
	cmp byte ptr ttt, '*' { Restore original color via ANSI EscSeq if needed }
	je @7
	mov ah, 9
	mov dx, offset ttt[25]
	int 21h
@7:     mov al, 0b3h  { '│' } { The line ends by this, too }
	int 29h
	mov al, 0 { The return value is FALSE }

{ END OF ALIGN }

@9:     or al, al
	jz lld  { Line not empty, print next line. }

lle:    { Print border and footer, then exit to DOS with EXIT_SUCCESS (0). }
	push $D9C0  { Print border '└┘'. }
	call PrintLine
	push ds
	push offset footermsg
	call Header
	mov ah, 3Eh      { close(F); }
	mov bx, qqq.han
	int 21h
llf:end; { ASM Statement. Good bye, reader. }
end.
